From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Wed, 1 Mar 2017 11:00:16 +0100
Subject: [PATCH] media: coda: make cache pagemerge configuration configurable
 from userspace

The page merge setting has a significant influence on the burst length
of read accesses. Until the effects of the page merge bitfield are
better known, allow to configure it from userspace.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/media/platform/coda/coda-bit.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/coda/coda-bit.c b/drivers/media/platform/coda/coda-bit.c
index fe10b2d0fa71..cae4753f276f 100644
--- a/drivers/media/platform/coda/coda-bit.c
+++ b/drivers/media/platform/coda/coda-bit.c
@@ -42,6 +42,10 @@
 #define CODA_DEFAULT_GAMMA	4096
 #define CODA9_DEFAULT_GAMMA	24576	/* 0.75 * 32768 */
 
+static int cache_pagemerge = 2;
+module_param(cache_pagemerge, int, 0644);
+MODULE_PARM_DESC(cache_pagemerge, "2D cache page merge setting");
+
 static int disable_encoder_cache = 0;
 module_param(disable_encoder_cache, int, 0644);
 MODULE_PARM_DESC(disable_encoder_cache, "Disable 2D cache for encoder");
@@ -830,7 +834,8 @@ static void coda9_set_frame_cache(struct coda_ctx *ctx, u32 fourcc)
 		 * artifacts in the interleaved chroma plane when decoding.
 		 */
 		cache_size = 0x02440224;
-		cache_config = 2 << CODA9_CACHE_PAGEMERGE_OFFSET;
+		cache_pagemerge &= 3;
+		cache_config = cache_pagemerge << CODA9_CACHE_PAGEMERGE_OFFSET;
 	}
 	coda_write(ctx->dev, cache_size, CODA9_CMD_SET_FRAME_CACHE_SIZE);
 	if (fourcc == V4L2_PIX_FMT_NV12 || fourcc == V4L2_PIX_FMT_YUYV) {
